Adds IT that verifies scans see data written by concurrent writers

keith-turner opened the pull request:
The IT ensures that when there are concurrent writes, scans, and table operations that scans always see any data written before the scan started.

keith-turner commented:
Created this IT for testing #3634, however realized its good for 2.1.X and #3573 to stress read patterns the metadata table relies on, so decided to create it in 2.1 and merge it forward to elasticity.

dlmarion commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -0,0 +1,492 @@
+
+          tctx.client.tableOperations().addSplits(tctx.table, splits);
+          numSplits += splitsToAdd;
+        }
```
should we add merging?

keith-turner replied:
yeah that would be good to add.  maybe bulk import too

keith-turner replied:
added merge and other stuff in 7cbc635

dlmarion commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -0,0 +1,492 @@
+      var tableOpsTask = executor.submit(new TableOpsTask(testContext));
+
+      // let the concurrent mayhem run for a bit
+      Thread.sleep(60000);
```
Is there any value to letting this run longer?

keith-turner replied:
Yeah I think there is value in running it longer, but maybe not in the IT.  We definitely want to know if does not work on a short run.  I think I will add a main method  that allows running it.  I think it would be good to have as an IT and something that is easy to run for longer periods.

keith-turner replied:
Opened #3646

dlmarion commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -0,0 +1,492 @@
+              writer.addMutation(m);
+              stats.written += m.getUpdates().size();
+              dataAdded.add(m);
+            }
```
Is it possible that the BatchWriter flushes in this loop based on default configs? If so, a scan that starts during this loop will see more data than expected.

keith-turner replied:
Yeah that is ok, the scans do not care if they see extra data. They just require that they see all data that was successfully written before the scan started.

There is a very last scan done after all the writers have stopped, that scan does not tolerate extra data and should throw an exception if it sees it.

dlmarion commented:
This test is a great idea

keith-turner commented:
I added merge, bulk import, compaction filtering, and batch scan to the test in 7cbc635.  I also added some debugging tips based on a problem I ran into where bulk importing deletes was causing the test to fail.  Spent hours trying to track down the cause and finally developed a technique that made tracking the problem down really quick.  I documented the technique [in the code](https://github.com/apache/accumulo/blob/7cbc6356fecb569a713094dba779f5e54c9301c9/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L93-L108) and left a [utility method](https://github.com/apache/accumulo/blob/7cbc6356fecb569a713094dba779f5e54c9301c9/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L668-L696) related to the technique in the test.

The problem I ran into is documented [here in the test](https://github.com/apache/accumulo/blob/7cbc6356fecb569a713094dba779f5e54c9301c9/test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java#L488-L496).  The problem is caused by the fact that major compactions no longer force a flush, this is a change made some time in the past, trying to find an issue for that.  Not flushing before a full major compaction is ok when only doing live ingest or only doing bulk import, but for a mix of bulk import and live ingest not flushing before full major compaction has a consequence as I discovered in the test.

EdColeman commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -0,0 +1,697 @@
+        return 0;
+      }
+
+      Path bulkDir = new Path(tctx.tmpDir + "/" + "bulkimport_" + nextLongAbs(random));
```
```suggestion
      Path bulkDir = new Path(tctx.tmpDir + "/bulkimport_" + nextLongAbs(random));
```

EdColeman commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/DataFileValue.java:
```diff
@@ -105,7 +105,7 @@ public int hashCode() {
   @Override
   public String toString() {
-    return size + " " + numEntries;
+    return size + " " + numEntries + " " + time;
```
I realize that this adds to the existing pattern - but would it be worth it to use a more standardized toString format that includes the field names? Otherwise, you need to read the code to understand that '12 1000 1234` is 'size: 12, numEntries 1000, time 1234'

keith-turner replied:
It would be good to improve the toString.  I am thinking maybe I should not make this change in 2.1 because I didn't analyze how this toString method was used. I made the change when trying to figure out the problem with bulk importing delete keys.  I'll either improve it or not change it after looking into how its used.

keith-turner replied:
That class is used in a huge number of places.  I decided to err on the side of caution and not change its toString rather try to analyze everywhere its used.

EdColeman reviewed (approved):
Minor suggestion on the toString, but that can be ignored if there is a reason not to use a standard toString format,

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java:
```diff
@@ -0,0 +1,697 @@
+ * This test verifies that scans will always see data written before the scan started even when
+ * there are concurrent scans, writes, and table operations running.
+ */
+@Tag(SUNNY_DAY)
```
I don't think it makes sense to add this to the sunny tests, as it's not testing base functionality, but it's testing more complicated edge cases that can be left in the larger test suite.
